Add FilterDisplayManager

cea2aj opened the pull request:
Add `FilterDisplayManager` for showing and hiding the filters based on the screen width and the view/hide filter buttons

A `PageView` is added to context through the `PageViewContextProvider`, and any component or page can access or modify it. There is a `FilterDisplayManager` which renders the filters based on the page view. Filters and dividers are passed to it using React props children. On desktop, it displays the filters, and on mobile it collapses them. If the `ViewFiltersButton` is pressed, the filters will expand to fill the whole page.

The mocks show that the filter dividers in the mobile expanded filter view goes to the edge of the screen. In order to accomplish this, I modify the styling of any `Facets`, `StaticFilters`, or `Divider` components inside the `FilterDisplayManager` so that the divider takes up the full width. Unfortunately this means that those components inside the `CollapsibleFiltersDrawer` are forced to use the 'assign' CompositionMethod. It is worth considering alternative approaches in the future on how we can override styling out-of-the-box and also allow the user to provide any kind of css `CompositionMethod`. Also, the Applied Filters need different styling when they are used in the universal results component. This is another use case where the ability to apply custom styling within the search starter and cleanly merge it with the user defined styling would be helpful. For now, I opted to copy the styling into the `SectionHeader` and make it customizable that way. This does introduce a little but of duplication to the styling, however I think that is preferable to modifying to the component on a vertical and having unintended conflicts on the universal page.

J=SLAP-1774
TEST=manual

Test on both mobile and and desktop and various screen sizes. See that on desktop the filters are visible. On mobile, the filters hide. I can press the button to see the filters, and scroll if needed. I can click the close button to go back to the mobile view. Resizing the window from the FiltersVisibleMobile to Desktop will properly update to the Desktop view.

nmanu1 commented on src/components/CollapsibleFilterDrawer.tsx:
```diff
@@ -0,0 +1,71 @@
+  if (pageView === PageView.Desktop) {
+    return <div className={cssClasses.container___desktop}>{children}</div>;
+  } else if (pageView === PageView.MobileFiltersExpanded) {
+    const childrenWithWithUpdatedDividers = recursivelyMapChildren(children, child => {
```
WithWith

oshi97 commented on src/hooks/usePageView.tsx:
```diff
@@ -0,0 +1,44 @@
+      }
+    }
+
+    window.addEventListener('resize', resizeListener);
```
is it possible to use resize observer here? not sure how hard it would be

cea2aj replied:
sure, that seems like more of a proper way of doing it

cea2aj replied:
After further investigation it looks like resize observer is used for elements whereas the widow resize event is more commonly used for changes to the whole window, so we're sticking with the window resize event

oshi97 commented on src/components/CollapsibleFilterDrawer.tsx:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Responsible for managing the CollapsibleFilterDrawer based on the current PageView
+ */
+export default function CollapsibleFilterDrawer({
```
what is a "drawer"?

oshi97 replied:
sure

yen-tt commented on src/components/CollapsibleFilterContainer.tsx:
```diff
@@ -0,0 +1,71 @@
+    });
+    return (
+      <div className={cssClasses.container___mobileFiltersExpanded}>
+        <CollapseFiltersButton onClick={() => setPageView(PageView.MobileFiltersCollapsed)} />
```
Can we directly add the button here instead of creating a new React component for it since it doesn't seem to be use anywhere else? Or is there a way to make a generic CollapsibleFiltersButton that can work for collapse and expand behavior with whatever onClick function and css pass in?

tmeyer2115 replied:
I think moving the `CollapseFiltersButton` into here makes sense. It won't be used elsewhere. If we were to add a `renderToggle` method that creates the button in this function, we could get rid of the `onClick` prop. The `onClick` prop feels a bit strange to me anyway. What happens when you click the button isn't really configurable, it should always collapse or expand.

cea2aj replied:
I'll move the button into the CollapsibleFilterContainer. I actually think it's simplest to get rid of the function to render the button and instead render it right inside the component

yen-tt commented on src/components/CollapsibleFilterContainer.tsx:
```diff
@@ -0,0 +1,71 @@
+import StaticFilters, { Divider } from "./StaticFilters";
+import recursivelyMapChildren from "./utils/recursivelyMapChildren";
+
+interface CollapsibleFilterContainerCssClasses {
```
should this extends 1CollapseFiltersButtonCssClasses`?

cea2aj replied:
after moving the button this interface is no longer needed

tmeyer2115 commented:
> 

What's meant by "Applied Filters need different styling when they are used in the universal results component"? I'm not aware of us displaying Applied Filters on the Universal page. In the Figma mocks, Michael has Filters and Applied Filters under the 'All' tab, but I think that's just a mistake. He didn't update the tab to be "Vertical".

tmeyer2115 commented on src/components/CollapsibleFilterContainer.tsx:
```diff
@@ -0,0 +1,71 @@
+};
+
+/**
+ * Responsible for managing the CollapsibleFilterContainer based on the current PageView
```
Can we update the JSDocs here? This reads a bit like a tautology: "The CollapsibleFilterContainer is responsible for managing the CollapsibleFilterContainer"

tmeyer2115 commented on src/pages/EventsPage.tsx:
```diff
@@ -28,40 +31,46 @@ const staticFiltersConfig = [{
+          <div className='flex'>
+            <ResultsCount />
+            {pageView === PageView.MobileFiltersCollapsed && 
+              <ExpandFiltersButton onClick={() => setPageView(PageView.MobileFiltersExpanded)}/>}
```
Is there any way we could include the expand functionality in the `CollapsibleFilterContainer`? It would mean we don't need to duplicate this code in every vertical page. I think it's also better encapsulation. One Component handles the expansion and collapse of the filters. Now, the logic for this is distributed across multiple areas of the code.

cea2aj replied:
Yes, it's possible, but there are a few reasons that lead me to this approach. The CollapsibleFilterContainer is in a different spot on the DOM that the expand filters button. When the  `ExpandFiltersButton` is it's own component, it's free to be placed anywhere on the page. I think it would be odd to have the `ExpandFiltersButton` appear on the page next to `ResultsCount` without the react code indicating that. However if we did want to go with that approach, the entire page would need to be managed by a component with both the `CollapsibleFiltersContainer` and the rest of the page inside of it. Perhaps we could have a `ResultsContainer` to contain the rest of the results page?

Alternatively we could pass the `pageView` to the `ExpandFiltersButton` so that it only renders itself on mobile with filters collapsed, which would let us get rid of that pageView conditional above it.

cea2aj commented:
> > 
> 
> What's meant by "Applied Filters need different styling when they are used in the universal results component"? I'm not aware of us displaying Applied Filters on the Universal page. In the Figma mocks, Michael has Filters and Applied Filters under the 'All' tab, but I think that's just a mistake. He didn't update the tab to be "Vertical".

When the filters are on the universal page next to the section header name, the spacing needs to be different. Applied filters on universal is functionality that we added to the sample app a while ago to match Theme 1.0. When we styled applied filters on vertical, I adjusted the padding so that it worked on universal as well. But you're right, I don't see the mocks for it on universal. I can double check with Michael if we are supposed to display them

cea2aj commented on src/components/CollapsibleFilterContainer.tsx:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Responsible for managing the CollapsibleFilterContainer based on the current PageView
+ */
+export default function CollapsibleFilterContainer({
```
Per our discussion, I have been thinking of different names for this component. What are your thoughts on `FilterDisplayManager` or `FilterViewManager`? Or maybe even `FilterPresenter`

yen-tt commented on src/hooks/usePageView.tsx:
```diff
@@ -0,0 +1,44 @@
+    isMobile() ? PageView.FiltersHiddenMobile : PageView.Desktop);
+
+  function isMobile () {
+    return window.innerWidth <= 768;
```
I'm wondering if it make sense to create a pageViewContext instead of a hook, and wrap it as a provider in the App component or just call useContext in whatever component/page that requires it without passing the pageView/setPageView field around?

cea2aj replied:
I think that's worth considering since it would let us pass around less data/functions and make it easy for any component to access it. I'll leave it up to @tmeyer2115 if we want to try attempt that now or in a future item

cea2aj replied:
Some more info from the React docs: "Context is primarily used when some data needs to be accessible by many components at different nesting levels. Apply it sparingly because it makes component reuse more difficult."

cea2aj replied:
Updated to use context!
